contracts/CarryPublicTokenCrowdsale.sol

```diff
@@ -0,0 +1,141 @@
+// The Carry token and the tokensale contracts
+// Copyright (C) 2018 Carry Protocol
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity ^0.4.24;
+
+import "openzeppelin-solidity/contracts/crowdsale/validation/CappedCrowdsale.sol";
+import "openzeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "./CarryToken.sol";
+
+/**
+ * @title CarryTokenCrowdsale
+ * @dev The common base contract for both sales: the Carry token presale,
+ * and the Carry token public crowdsale.
+ */
+contract CarryPublicTokenCrowdsale is CappedCrowdsale, Pausable {
+    using SafeMath for uint256;
+
+    // Individual min purchases.
+    uint256 public individualMinPurchaseWei;
+
+    struct IndividualMaxCap {
+        uint256 timestamp;
+        uint256 maxWei;
+    }
+
+    // Individual max purchases differ by time.  The mapping keys are timestamps
+    // and values are weis an individual can purchase at most
+    // If the transaction is made later than a timestamp it can accept
+    // the corresponding cap at most.
+    //
+    // Where individualMaxCaps = [
+    //   IndividualMaxCap(1533081600000, 5 ether),
+    //   IndividualMaxCap(1533686400000, 10 ehter)
+    // ]
+    // If a transaction is made before 1533081600000 (2018-08-01 sharp UTC)
```

Comment:
Looks like ethereum timestamps are second-based not millisecond-based([doc](https://solidity.readthedocs.io/en/latest/units-and-global-variables.html?highlight=timestamp#block-and-transaction-properties)). Is there a reason millisecond-based timestamp was chosen instead?